lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +652,1142 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+                            CallSites.Arguments)) {
+            auto *CSV = FunctionArg == nullptr ? CSArg->first :
+                                                 FunctionArg->first;
+            RegisterState RSFArg = FunctionArg == nullptr ?
+                                     RegisterState::Maybe :
+                                     FunctionArg->second;
+            RegisterState RSCSArg = CSArg == nullptr ? RegisterState::Maybe :
+                                                       CSArg->second;
+
+            Summary.ABIResults.Arguments[CSV] = ABIAnalyses::combine(RSFArg,
+                                                                     RSCSArg);
+          }
+        }
+      }
+    }
+  }
+}
+
+static UpcastablePointer<model::Type>
+buildPrototype(GeneratedCodeBasicInfo &GCBI,
+               model::Binary &Binary,
+               const FunctionSummary &Summary,
+               const model::BasicBlock &Block) {
+  using namespace model;
+  using RegisterState = model::RegisterState::Values;
+
+  auto NewType = makeType<RawFunctionType>();
+  auto &CallType = *llvm::cast<RawFunctionType>(NewType.get());
+  {
+    auto ArgumentsInserter = CallType.Arguments.batch_insert();
+    auto ReturnValuesInserter = CallType.ReturnValues.batch_insert();
+
+    bool Found = false;
+    for (const auto &[PC, CallSites] : Summary.ABIResults.CallSites) {
+      if (PC != Block.Start)
+        continue;
+
+      revng_assert(!Found);
+      Found = true;
+
+      for (const auto &[Arg, RV] :
+           zipmap_range(CallSites.Arguments, CallSites.ReturnValues)) {
+        auto *CSV = Arg == nullptr ? RV->first : Arg->first;
+        RegisterState RSArg = Arg == nullptr ? RegisterState::Maybe :
+                                               Arg->second;
+        RegisterState RSRV = RV == nullptr ? RegisterState::Maybe : RV->second;
+
+        auto RegisterID = ABIRegister::fromCSVName(CSV->getName(), GCBI.arch());
+        if (RegisterID == Register::Invalid || CSV == GCBI.spReg())
+          continue;
+
+        auto *CSVType = CSV->getType()->getPointerElementType();
+        auto CSVSize = CSVType->getIntegerBitWidth() / 8;
+        NamedTypedRegister TR(RegisterID);
+        TR.Type = {
+          Binary.getPrimitiveType(PrimitiveTypeKind::Generic, CSVSize), {}
+        };
+
+        if (model::RegisterState::shouldEmit(RSArg))
+          ArgumentsInserter.insert(TR);
+
+        if (model::RegisterState::shouldEmit(RSRV))
+          ReturnValuesInserter.insert(TR);
+      }
+    }
+    revng_assert(Found);
+
+    CallType.PreservedRegisters = {};
+    CallType.FinalStackOffset = 0;
+  }
+
+  return NewType;
+}
+
+static void
+finalizeModel(GeneratedCodeBasicInfo &GCBI,
+              const std::vector<CFEP> &Functions,
+              const std::vector<llvm::GlobalVariable *> &ABIRegisters,
+              const FunctionAnalysisResults &Properties,
+              model::Binary &Binary) {
```

Comment:
Many of these arguments seem to be fields of `CFEPAnalyzer`, why not make a method out of it?

Reply:
Because IMHO, `finalizeModel` has nothing (or very little) to do with `CFEPAnalyzer`. Like the population of the model should have no relationship with the analysis of the CFEP, and it is something that could happen even past the lifetime of the instance of `CFEPAnalyzer`. Perhaps it could make sense if it was something like `finalizeCFG` (since there's `collectDirectCFG`), but, as far as I see things, `finalizeModel` should be more like a static function.

Reply:
Makes sense.